File: frontend/src/utils/cache.ts
// frontend/src/utils/cache.ts

interface CacheItem<T> {
  data: T;
  timestamp: number;
  expiresAt: number;
}

class CacheManager {
  private cache: Map<string, CacheItem<any>> = new Map();
  private defaultTTL = 5 * 60 * 1000; // 5 minutes default

  set<T>(key: string, data: T, ttlMs?: number): void {
    const now = Date.now();
    const ttl = ttlMs || this.defaultTTL;
    
    this.cache.set(key, {
      data,
      timestamp: now,
      expiresAt: now + ttl
    });
  }

  get<T>(key: string): T | null {
    const item = this.cache.get(key);
    
    if (!item) {
      return null;
    }

    // Check if expired
    if (Date.now() > item.expiresAt) {
      this.cache.delete(key);
      return null;
    }

    return item.data as T;
  }

  has(key: string): boolean {
    const item = this.cache.get(key);
    if (!item) return false;
    
    if (Date.now() > item.expiresAt) {
      this.cache.delete(key);
      return false;
    }
    
    return true;
  }

  clear(): void {
    this.cache.clear();
  }

  clearExpired(): void {
    const now = Date.now();
    for (const [key, item] of this.cache.entries()) {
      if (now > item.expiresAt) {
        this.cache.delete(key);
      }
    }
  }

  // Get age of cached item in milliseconds
  getAge(key: string): number | null {
    const item = this.cache.get(key);
    if (!item) return null;
    return Date.now() - item.timestamp;
  }
}

// Export a singleton instance
export const cache = new CacheManager();

// Cache keys generator
export const cacheKeys = {
  productInfo: (urls: string[]): string => `product-info:${urls.sort().join(',')}`,
  productDetails: (url: string): string => `product-details:${url}`,
  watchlist: (username: string): string => `watchlist:${username}`,
  userSettings: (username: string): string => `user-settings:${username}`
};

File: backend/cache.py
# backend/cache.py
import json
import os
import time
from typing import Optional, Any
from hashlib import md5

CACHE_DIR = "cache"
DEFAULT_TTL = 300  # 5 minutes

class CacheManager:
    def __init__(self, cache_dir: str = CACHE_DIR):
        self.cache_dir = cache_dir
        os.makedirs(cache_dir, exist_ok=True)
    
    def _get_cache_path(self, key: str) -> str:
        """Generate a safe filename from cache key"""
        safe_key = md5(key.encode()).hexdigest()
        return os.path.join(self.cache_dir, f"{safe_key}.json")
    
    def get(self, key: str) -> Optional[Any]:
        """Get cached value if it exists and hasn't expired"""
        cache_path = self._get_cache_path(key)
        
        if not os.path.exists(cache_path):
            return None
        
        try:
            with open(cache_path, 'r') as f:
                cache_data = json.load(f)
            
            # Check if expired
            if time.time() > cache_data['expires_at']:
                os.remove(cache_path)
                return None
            
            return cache_data['data']
        except Exception:
            return None
    
    def set(self, key: str, data: Any, ttl: int = DEFAULT_TTL):
        """Cache data with TTL in seconds"""
        cache_path = self._get_cache_path(key)
        
        cache_data = {
            'data': data,
            'timestamp': time.time(),
            'expires_at': time.time() + ttl,
            'key': key
        }
        
        with open(cache_path, 'w') as f:
            json.dump(cache_data, f)
    
    def delete(self, key: str):
        """Delete cached item"""
        cache_path = self._get_cache_path(key)
        if os.path.exists(cache_path):
            os.remove(cache_path)
    
    def clear_expired(self):
        """Clear all expired cache entries"""
        current_time = time.time()
        
        for filename in os.listdir(self.cache_dir):
            if filename.endswith('.json'):
                filepath = os.path.join(self.cache_dir, filename)
                try:
                    with open(filepath, 'r') as f:
                        cache_data = json.load(f)
                    
                    if current_time > cache_data.get('expires_at', 0):
                        os.remove(filepath)
                except Exception:
                    # Remove corrupted cache files
                    os.remove(filepath)

# Global cache instance
cache = CacheManager()